src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -1205,6 +1206,161 @@ static async System.Collections.Generic.IAsyncEnumerator<int> M(int value)
+                );
+        }
+
+        [Fact]
+        [WorkItem(31552, "https://github.com/dotnet/roslyn/issues/31552")]
+        public void AsyncIteratorReturningEnumerator_WithAwaitAndThrow()
+        {
+            string source = @"
+class C
+{
+    async System.Collections.Generic.IAsyncEnumerator<int> M()
+    {
+        await System.Threading.Tasks.Task.CompletedTask;
+        throw new System.NotImplementedException();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyEmitDiagnostics(
+                // (4,60): error CS8419: The body of an async-iterator method must contain a 'yield' statement.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.ERR_PossibleAsyncIteratorWithoutYield, "M").WithArguments("System.Collections.Generic.IAsyncEnumerator<int>").WithLocation(4, 60)
+                );
+        }
+
+        [Fact]
+        [WorkItem(31552, "https://github.com/dotnet/roslyn/issues/31552")]
+        public void AsyncIteratorReturningEnumerator_WithThrow_WithAwaitInLambda()
+        {
+            string source = @"
+class C
+{
+    async System.Collections.Generic.IAsyncEnumerator<int> M()
+    {
+        System.Func<System.Threading.Tasks.Task> lambda = async () => { await System.Threading.Tasks.Task.CompletedTask; };
+        throw new System.NotImplementedException();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(source);
+            comp.VerifyDiagnostics(
+                // (4,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(4, 60)
+                );
+            comp.VerifyEmitDiagnostics(
+                // (4,60): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(4, 60),
+                // (4,60): error CS8420: The body of an async-iterator method must contain a 'yield' statement. Consider removing `async` from the method declaration.
+                //     async System.Collections.Generic.IAsyncEnumerator<int> M()
+                Diagnostic(ErrorCode.ERR_PossibleAsyncIteratorWithoutYieldOrAwait, "M").WithArguments("System.Collections.Generic.IAsyncEnumerator<int>").WithLocation(4, 60)
```

Comment:
How come this is an error rather than a warning?

Reply:
@alrz Imagine that the `IAsyncEnumerable` type was tagged to make it a task-like attribute. We don't know whether to emit a state machine for async method, or for async-iterator method.
If we can't compile, then the diagnostic must be an error.